lib/_stream_writable.js

```diff
@@ -590,6 +590,13 @@ Writable.prototype.end = function(chunk, encoding, cb) {
   // Ignore unnecessary end() calls.
   if (!state.ending)
     endWritable(this, state, cb);
+  else if (typeof cb === 'function') {
+    if (!state.finished) {
+      this.once('finish', cb);
+    } else {
+      cb();
```

Comment:
Should the error be used here?

Reply:
It's not used in stream. I don't think we need the error if we do it like this?

Reply:
Well, it `'finish'` event has been already emitted now, so this should not be called at all or called with an error.

Reply:
ah, that makes sense. Should I rename the error to something along the lines of `ERR_STREAM_FINISHED`?

Reply:
yes, makes sense.